Automattic/wp-calypso
# Separate post likes and likers state

Fixes #84213

## Proposed Changes

* This PR creates a dedicated `QueryPostLikers` component to query for likes information, and splits up the state handling for basic post likes information, specifically the number of likes and whether the current user likes the post, and the likers information, specifically which users liked the post.
* The reason for making this change is that the bug reported in #84213 occurs because we end up making two API requests: one when we show the popover, and one to like the post when the user clicks, but we end up overwriting the number of likes and liked state when the response from the popover comes back.
* To avoid further race conditions, the solution in this PR is to update the `iLike` and `found` state if and only if we're performing a like-related action, and to update the `likes` state if and only if we're specifically updating the `likes` data. This avoids cases where slow calls to fetch likes data can overwrite likes performed on the client, as in the original issue.

## Testing Instructions

### Testing procedure

* Open the Reader and find a post card with multiple existing likes
* Mouse over the like star and immediately click on it
* [Optional extension - h/t @roo2] Click on the like star again in short succession

### Test cases

* Run WordPress.com _without_ this patch.
* Follow the testing procedure above, and verify that you see the like state and like count behave as reported in the original bug -- the end like state should be _unliked_ (i.e. not filled), and the like count should toggle unexpectedly
  - This is a timing issue/race condition, but we want to verify that you can reproduce the problem
  - If you are having trouble reproducing the issue, try throttling your network connection to see if that makes it easier to reproduce. This is probably easiest to do via your browser dev tools.
* Run this branch locally or via Calypso.live
* Follow the testing procedure above.
* Verify that the like state updates immediately so the star is filled and the liker count is updated, and that the popover shows your username without any changes to the like count or star fill state

## Pre-merge Checklist



- [N/A] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [N/A] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [N/A] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [N/A] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
